argh/src/lib.rs

```diff
@@ -443,6 +443,11 @@ pub trait TopLevelCommand: FromArgs {}
 pub trait SubCommands: FromArgs {
     /// Info for the commands.
     const COMMANDS: &'static [&'static CommandInfo];
+
+    /// Get a list of commands that are discovered at runtime.
+    fn dynamic_commands() -> &'static [&'static dyn DynamicCommand<Self>] {
```

Comment:
Materializing the command list can in theory be expensive. We don't want it to happen more than once usually. OnceCell accomplishes that, and 'static tends to encourage that kind of solution. Otherwise we either have to warn users that we can call their materialization process many times and they should be aware of that for performance reasons, or take extra measures to ensure we only call it once internally.

Reply:
how often would we be calling `DynamicSubCommands::commands()`? I would have expected that we'd only call it once for each recursive `FromArgs::from_args()`, generating a help message, or `FromArgs::redact_arg_values()`. Do you know if there's more?

While I'm not opposed to this returning `&'static ...`, It'd be nice to not force people to have to use something like OnceCell since we don't really have an easy-to-use implementation that's just in the standard library (this is apparently coming at some point in time with https://github.com/rust-lang/rust/issues/74465 though).